File: 0009-palindrome-number/0009-palindrome-number.scala
import scala.annotation.tailrec

object Solution {
    def isPalindrome(x: Int): Boolean = {
        if (x < 0 || x % 10 == 0 && x != 0) {
            false
        } else {
            @tailrec
            def reverse(currentX: Int, currentReversed: Int): Int = {
                if (currentX == 0) {
                    currentReversed
                } else {
                    val lastDigit = currentX % 10
                    val nextReversed = (currentReversed * 10) + lastDigit
                    reverse(currentX / 10, nextReversed)
                }
            }
            reverse(x, 0) == x
        }
        
    }
}

File: 0009-palindrome-number/0009-palindrome-number.go
func isPalindrome(x int) bool {

    if x < 0 || (x%10 == 0 && x != 0) {
            return false
    }

    reversed := 0
    orig := x

    for x > 0 {
        reversed = (reversed * 10) + (x % 10)
        x /= 10
    }

    return orig == reversed
    
}